docs/spec.rst

```diff
@@ -0,0 +1,501 @@
+method return values, e.g., when methods return field values. We
+therefore likewise only expect non-null guarantees to hold for instance
+(static, respectively) fields and method results by the time their
+declaring class’s constructor (static initializer, respectively) has
+finished (similar to when final fields are guaranteed to be
+initialized).
+
+Note this semantics does allow for situations in which null values may
+be observable in “non-null” fields and method returns, namely while
+objects are under construction. We simply encourage API owners to
+minimize these cases for non-private (both static and instance) fields
+and methods, which typically involves not “leaking” object references
+outside an API until they’re fully constructed. Some tools may attempt
+to identify such “leaks” and may attempt to ensure proper field
+initialization as defined here during object construction.
+
+Examples
+~~~~~~~~
+
+As an example, let’s consider a fragment of Guava’s ``ImmutableMap``:
+
+.. code:: java
+
+   @DefaultNonNull
+   public class ImmutableMap<K, V> implements Map<K, V> {
+     public static <K, V> ImmutableMap<K, V> of(K key, V value);
+     public @Nullable V get(@Nullable Object key);
+   }
+
+Because of the use of ``@DefaultNonNull``, every type use in this
+class’s declaration is fixed to either nullable or non-null (including
+type-variable uses, since their type parameters are considered
+implicitly bounded by non-null ``Object``).
+
+-  What’s ``of()``\ ‘s formal type parameters’ nullability? non-null
+   (from ``K`` and ``V``\ ’s bounds, which are determined implicitly by
+   ``@DefaultNonNull``)
+-  What’s ``get()``\ ’s return value’s nullability? nullable (from its
+   explicit annotation)
+-  It is a mismatch to refer to
+   ``ImmutableMap<@Nullable String, Object>`` because
+   ``@Nullable String`` is outside of ``K``\ ’s bounds.
+
+To illustrate wildcards, consider a method return type
+``ImmutableMap<? extends @Nullable String, ?>`` with no defaulting
+annotation in scope:
+
+-  What’s the map’s nullability? unspecified (since no defaulting
+   annotation is in scope)
+-  What’s the nullability of the map’s first wildcard? non-null. This is
+   because the wildcard’s explicit bound, ``@Nullable String``, is
+   intersected with the wildcarded type parameter ``K``\ ’s bound,
+   non-null ``Object``. In other words, the explicit ``@Nullable`` here
+   is ignored and might justify a warning. That warning can be avoided
+   by writing ``? extends String`` without changing the nullability:
+   ``String``, being not explicitly annotated, has unspecified
+   nullability, but that’s still superseded by the wildcarded type
+   parameter’s bound, non-null ``Object``.
+-  What’s the nullability of the map’s second wildcard? Also non-null,
+   because the wildcard inherits that bound from the bound of ``V`` in
+   ``ImmutableMap``.
+
+As another example, ``java.util.function.Function`` would be declared as
```

Comment:
maybe we should use guava's Function interface here?

Reply:
Done.